File: backend/api/llm.py
from __future__ import annotations

from typing import Any, Dict, List

from fastapi import APIRouter
from pydantic import BaseModel

from ui.utils.llm_config import (
    configured_model_options,
    default_llm_profiles,
    normalize_llm_profiles,
    parse_selected_model,
    test_connectivity,
)

router = APIRouter()


class TestConnectivityRequest(BaseModel):
    provider: str
    api_key: str
    api_base: str = ""
    model_name: str


class NormalizeRequest(BaseModel):
    profiles: Dict[str, Dict[str, Any]]


@router.get("/defaults")
def get_defaults() -> Dict[str, Any]:
    return {"profiles": default_llm_profiles()}


@router.post("/test")
def test_connection(body: TestConnectivityRequest) -> Dict[str, Any]:
    ok, models, error = test_connectivity(
        body.provider, body.api_key, body.api_base, body.model_name
    )
    return {"ok": ok, "models": models, "error": error}


@router.post("/normalize")
def normalize(body: NormalizeRequest) -> Dict[str, Any]:
    return {"profiles": normalize_llm_profiles(body.profiles)}


@router.post("/options")
def options(body: NormalizeRequest) -> Dict[str, Any]:
    opts = configured_model_options(body.profiles)
    return {"options": [{"key": k, "label": v} for k, v in opts]}


class ParseModelRequest(BaseModel):
    key: str


@router.post("/parse")
def parse(body: ParseModelRequest) -> Dict[str, Any]:
    provider, model = parse_selected_model(body.key)
    return {"provider": provider, "model": model}

File: frontend/src/api/llm.ts
import { api } from "./client";
import type { LlmProfiles } from "@/types";

export async function getDefaults(): Promise<{ profiles: LlmProfiles }> {
  const { data } = await api.get("/llm/defaults");
  return data;
}

export async function testConnectivity(payload: {
  provider: string;
  api_key: string;
  api_base: string;
  model_name: string;
}): Promise<{ ok: boolean; models: string[]; error: string }> {
  const { data } = await api.post("/llm/test", payload);
  return data;
}

export async function normalize(profiles: LlmProfiles): Promise<LlmProfiles> {
  const { data } = await api.post("/llm/normalize", { profiles });
  return data.profiles;
}

export async function getOptions(
  profiles: LlmProfiles
): Promise<{ key: string; label: string }[]> {
  const { data } = await api.post("/llm/options", { profiles });
  return data.options;
}

export async function parseModel(
  key: string
): Promise<{ provider: string; model: string }> {
  const { data } = await api.post("/llm/parse", { key });
  return data;
}
